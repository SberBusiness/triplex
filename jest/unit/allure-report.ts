import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
