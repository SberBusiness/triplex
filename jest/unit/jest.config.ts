import type {Config} from '@jest/types';

// Конфигурация Jest для запуска unit тесов.
const config: Config.InitialOptions = {
  preset: 'ts-jest',
  rootDir: '../../',
  roots: [
    '<rootDir>/src/',
    '<rootDir>/tools/',
  ],
  setupFilesAfterEnv: [
    '<rootDir>/jest/jest.setup.ts',
    '<rootDir>/jest/unit/allure-report.ts',
  ],
  moduleDirectories: [
    'node_modules',
    'src'
  ],
  moduleNameMapper: {
    "@jest/(.*)": "<rootDir>/jest/$1.ts",
    "^@sberbusiness/triplex/(.*)": "<rootDir>/src/$1"
  },
  verbose: true,
};

export default config;