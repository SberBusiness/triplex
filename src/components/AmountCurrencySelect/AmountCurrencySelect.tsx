import React from 'react';
import {AmountBaseInput} from '../protected/AmountBaseInput/AmountBaseInput';
import {IInputProps} from '@sberbusiness/triplex/components/Input/Input';
import {uniqueId} from '@sberbusiness/triplex/utils/uniqueId';
import {CurrencySrvIcon20} from '@sberbusiness/icons/CurrencySrvIcon20';
import {TestIds} from '@sberbusiness/triplex/dataTestIds/dataTestIds';
import {TestProps} from '@sberbusiness/triplex/types/CoreTypes';
import {ISelectOption, ISelectProps} from '@sberbusiness/triplex/components/Select/Select';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {ISelectExtendedDropdownProvideProps, ISelectExtendedTargetProvideProps, SelectExtended} from '../SelectExtended/SelectExtended';
import {DropdownListContext} from '@sberbusiness/triplex/components/Dropdown/DropdownListContext';

/** Свойства AmountCurrencySelect. */
export interface IAmountCurrencySelectProps extends Omit<IInputProps, 'type' | 'maxLength' | 'onChange' | 'onSelect'>, TestProps {
    /** Значение.  */
    value: string;
    /** Максимальное количество знаков перед запятой. */
    maxIntegerDigits?: number;
    /** Количество знаков после запятой. */
    fractionDigits?: number;
    /** Выбранная валюта. */
    currency: ISelectOption | null;
    /** Список валют. */
    currencyOptions: ISelectOption[];
    /** Обработчик выбора элемента списка. */
    onSelect: (item: ISelectOption) => void;
    /** Флаг для отображения спиннера. */
    loading?: boolean;
    /** Свойства компонента выбора валюты. */
    currencyProps?: ISelectProps;
    /** Обработчик изменения значения. */
    onChange: (value: string) => void;
}

/** Состояния AmountCurrencySelect. */
export interface IAmountCurrencySelectState {
    activeDescendant?: string;
}

/**
 * Компонент ввода суммы и выбора валюты.
 *
 * @deprecated Используйте компонент InputGroup.
 */
export class AmountCurrencySelect extends React.Component<IAmountCurrencySelectProps, IAmountCurrencySelectState> {
    public static displayName = 'AmountCurrencySelect';

    public state = {
        activeDescendant: undefined,
    };

    public static defaultProps = {
        fractionDigits: 2,
    };

    private instanceId = uniqueId();

    private readonly targetRef: React.RefObject<HTMLDivElement>;

    constructor(props: IAmountCurrencySelectProps) {
        super(props);

        this.targetRef = React.createRef();
    }

    public render(): JSX.Element {
        const {
            className,
            currency,
            currencyOptions,
            onSelect,
            loading,
            currencyProps,
            'data-test-id': dataTestId,
            ...restProps
        } = this.props;
        const {tabIndex, disabled, error} = restProps;
        const {activeDescendant} = this.state;
        const currencyTabIndex = typeof tabIndex === 'number' ? tabIndex + 1 : undefined;

        const renderTarget = (props: ISelectExtendedTargetProvideProps) => (
            <SelectExtended.Target
                className={classnames('cssClass[currencySelectExtendedTarget]', {
                    'cssClass[currencySelectExtendedTargetOpened]': props.opened,
                })}
                label={currency ? currency.label : ''}
                disabled={disabled || currencyProps?.disabled}
                error={error}
                loading={loading}
                placeholder={<CurrencySrvIcon20 className="cssClass[currencyIcon]" />}
                tabIndex={currencyTabIndex}
                title={currencyProps?.title}
                role="combobox"
                aria-controls={this.instanceId}
                aria-activedescendant={activeDescendant}
                ref={this.targetRef}
                {...props}
            />
        );

        return (
            <div className={classnames('cssClass[amountCurrencySelect]', className)} data-test-id={dataTestId}>
                <AmountBaseInput {...restProps} data-test-id={dataTestId && `${dataTestId}${TestIds.AmountCurrencySelect.input}`} />

                <SelectExtended
                    className="cssClass[currencySelectExtended]"
                    data-test-id={dataTestId && `${dataTestId}${TestIds.AmountCurrencySelect.select}`}
                    renderTarget={renderTarget}
                    closeOnTab
                >
                    {({opened, setOpened, dropdownRef, ...restDropdownProps}: ISelectExtendedDropdownProvideProps) => (
                        <DropdownListContext.Provider value={{activeDescendant, setActiveDescendant: this.setActiveDescendant}}>
                            <SelectExtended.Dropdown
                                {...restDropdownProps}
                                className="cssClass[currencySelectExtendedDropdown]"
                                data-test-id={dataTestId && `${dataTestId}${TestIds.AmountCurrencySelect.dropdown}`}
                                opened={opened}
                                setOpened={setOpened}
                                targetRef={this.targetRef}
                                forwardedRef={dropdownRef}
                                fixedWidth={true}
                            >
                                <SelectExtended.Dropdown.List id={this.instanceId} dropdownOpened={opened}>
                                    {currencyOptions.map((option) => (
                                        <SelectExtended.Dropdown.List.Item
                                            {...option}
                                            key={option.value}
                                            id={option.value}
                                            className="cssClass[currencySelectExtendedDropdownListItem]"
                                            data-test-id={
                                                dataTestId &&
                                                `${dataTestId}${TestIds.AmountCurrencySelect.dropdown}${TestIds.Dropdown.listItem}`
                                            }
                                            selected={Boolean(currency && option.value === currency.value)}
                                            onSelect={() => {
                                                onSelect(option);
                                                setOpened(false);
                                            }}
                                        >
                                            {option.label}
                                        </SelectExtended.Dropdown.List.Item>
                                    ))}
                                </SelectExtended.Dropdown.List>
                            </SelectExtended.Dropdown>
                        </DropdownListContext.Provider>
                    )}
                </SelectExtended>
            </div>
        );
    }

    private setActiveDescendant = (id?: string) => {
        const {activeDescendant} = this.state;

        if (activeDescendant !== id) {
            this.setState({activeDescendant: id});
        }
    };
}
