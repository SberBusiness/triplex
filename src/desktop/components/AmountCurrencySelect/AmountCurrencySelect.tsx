import * as React from 'react';
import {uniqueId} from '@sberbusiness/triplex/common/utils/uniqueId';
import {CurrencySrvIcon20} from '@sberbusiness/icons/CurrencySrvIcon20';
import {TestIds} from '@sberbusiness/triplex/common/dataTestIds/dataTestIds';
import {TestProps} from '@sberbusiness/triplex/desktop/common/types/CoreTypes';
import {ISelectOption, ISelectProps} from '@sberbusiness/triplex/desktop/components/Select/Select';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import {NumberBaseInput} from '../protected/NumberBaseInput/NumberBaseInput';
import {INumberBaseInputProps} from '../protected/NumberBaseInput/types';
import {ISelectExtendedDropdownProvideProps, ISelectExtendedTargetProvideProps, SelectExtended} from '../SelectExtended/SelectExtended';
import {DropdownListContext} from '@sberbusiness/triplex/desktop/components/Dropdown/DropdownListContext';

/**
 * Свойства AmountCurrencySelect.
 *
 * @extends INumberBaseInputProps
 * @extends TestProps
 */
export interface IAmountCurrencySelectProps
    extends Omit<INumberBaseInputProps, 'onSelect' | 'valueFractionLength' | 'placeholderFractionLength'>,
        TestProps {
    /** Выбранная валюта.*/
    currency: ISelectOption | null;
    /** Список валют.*/
    currencyOptions: ISelectOption[];
    /** Обработчик выбора элемента списка.*/
    onSelect: (item: ISelectOption) => void;
    /** Флаг для отображения спиннера.*/
    loading?: boolean;
    /** Количество знаков после разделителя.*/
    fractionLength?: number;
    /** Свойства компонента выбора валюты.*/
    currencyProps?: ISelectProps;
}

/** Состояния AmountCurrencySelect. */
export interface IAmountCurrencySelectState {
    activeDescendant?: string;
}

/**
 * Компонент ввода суммы и выбора валюты.
 *
 * @deprecated Используйте компонент InputGroup.
 */
export class AmountCurrencySelect extends React.Component<IAmountCurrencySelectProps, IAmountCurrencySelectState> {
    public static displayName = 'AmountCurrencySelect';

    public state = {
        activeDescendant: undefined,
    };

    public static defaultProps = {
        fractionLength: 2,
    };

    private instanceId = uniqueId();

    private readonly targetRef: React.RefObject<HTMLDivElement>;

    constructor(props: IAmountCurrencySelectProps) {
        super(props);

        this.targetRef = React.createRef();
    }

    public render(): JSX.Element {
        const {
            className,
            error,
            disabled,
            currency,
            currencyOptions,
            onSelect,
            loading,
            fractionLength,
            tabIndex,
            currencyProps,
            'data-test-id': dataTestId,
            ...restProps
        } = this.props;
        const {activeDescendant} = this.state;
        const currencyTabIndex = typeof tabIndex === 'number' ? tabIndex + 1 : undefined;

        const renderTarget = (props: ISelectExtendedTargetProvideProps) => (
            <SelectExtended.Target
                className={classnames('cssClass[currencySelectExtendedTarget]', {
                    'cssClass[currencySelectExtendedTargetOpened]': props.opened,
                })}
                label={currency ? currency.label : ''}
                disabled={disabled || currencyProps?.disabled}
                error={error}
                loading={loading}
                placeholder={<CurrencySrvIcon20 className="cssClass[currencyIcon]" />}
                tabIndex={currencyTabIndex}
                title={currencyProps?.title}
                role="combobox"
                aria-controls={this.instanceId}
                aria-activedescendant={activeDescendant}
                ref={this.targetRef}
                {...props}
            />
        );

        return (
            <div className={classnames('cssClass[amountCurrencySelect]', className)} data-test-id={dataTestId}>
                <NumberBaseInput
                    {...restProps}
                    valueFractionLength={fractionLength}
                    placeholderFractionLength={fractionLength}
                    data-test-id={dataTestId && `${dataTestId}${TestIds.AmountCurrencySelect.input}`}
                    error={error}
                    disabled={disabled}
                    tabIndex={tabIndex}
                />

                <SelectExtended
                    className="cssClass[currencySelectExtended]"
                    renderTarget={renderTarget}
                    data-test-id={dataTestId && `${dataTestId}${TestIds.AmountCurrencySelect.select}`}
                >
                    {({
                        className: dropdownClassName,
                        opened,
                        setOpened,
                        dropdownRef,
                        ...restDropdownProps
                    }: ISelectExtendedDropdownProvideProps) => (
                        <DropdownListContext.Provider value={{activeDescendant, setActiveDescendant: this.setActiveDescendant}}>
                            <SelectExtended.Dropdown
                                {...restDropdownProps}
                                className={classnames('cssClass[currencySelectExtendedDropdown]', dropdownClassName)}
                                data-test-id={dataTestId && `${dataTestId}${TestIds.AmountCurrencySelect.dropdown}`}
                                opened={opened}
                                targetRef={this.targetRef}
                                forwardedRef={dropdownRef}
                                fixedWidth={true}
                            >
                                <SelectExtended.Dropdown.List id={this.instanceId} dropdownOpened={opened}>
                                    {currencyOptions.map((option) => (
                                        <SelectExtended.Dropdown.List.Item
                                            {...option}
                                            key={option.value}
                                            id={option.value}
                                            className="cssClass[currencySelectExtendedDropdownListItem]"
                                            data-test-id={
                                                dataTestId &&
                                                `${dataTestId}${TestIds.AmountCurrencySelect.dropdown}${TestIds.Dropdown.listItem}`
                                            }
                                            selected={Boolean(currency && option.value === currency.value)}
                                            onSelect={() => {
                                                onSelect(option);
                                                setOpened(false);
                                            }}
                                        >
                                            {option.label}
                                        </SelectExtended.Dropdown.List.Item>
                                    ))}
                                </SelectExtended.Dropdown.List>
                            </SelectExtended.Dropdown>
                        </DropdownListContext.Provider>
                    )}
                </SelectExtended>
            </div>
        );
    }

    private setActiveDescendant = (id?: string) => {
        const {activeDescendant} = this.state;

        if (activeDescendant !== id) {
            this.setState({activeDescendant: id});
        }
    };
}
