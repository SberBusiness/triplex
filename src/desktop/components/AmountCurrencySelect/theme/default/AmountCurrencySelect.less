@import '../../../../common/theme/default';

@ufs-theme-currency-select-width: 86px;

.amountCurrencySelect {
    display: flex;
    align-items: center;

    input {
        padding-right: @metric-12;
        border-radius: @ufs-theme-input-border-radius 0 0 @ufs-theme-input-border-radius;
        text-align: right;
    }

    .currencySelectExtendedTarget {
        border-radius: 0 @ufs-theme-input-border-radius @ufs-theme-input-border-radius 0;
    }

    .currencySelectExtendedTarget:focus .currencyIcon .service-fill,
    .currencySelectExtendedTargetOpened .currencyIcon .service-fill {
        fill: @color-asphalt-01;
    }

    .currencySelectExtended {
        position: relative;
        margin-left: -1px;
        flex: 0 0 @ufs-theme-currency-select-width;
        max-width: @ufs-theme-currency-select-width; // ie fix
    }

    .currencySelectExtendedDropdownListItem {
        overflow: hidden;
        text-overflow: ellipsis;
    }

    input:focus,
    .currencySelectExtended:focus {
        z-index: 2;
    }

    .currencyIcon {
        position: absolute;
        top: 0;
        bottom: 0;
        margin: auto;
    }
}
