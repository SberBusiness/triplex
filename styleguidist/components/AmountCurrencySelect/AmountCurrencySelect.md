```jsx
const [value, setValue] = React.useState('');
const [currency, setCurrency] = React.useState(null);

const currencyOptions = [
    {label: 'RUB', value: 'RUB'},
    {label: 'USD', value: 'USD'},
    {label: 'EUR', value: 'EUR'},
];

<AmountCurrencySelect
    value={value}
    currency={currency}
    currencyOptions={currencyOptions}
    currencyProps={{title: currency ? currency.label : undefined}}
    onChange={setValue}
    onSelect={setCurrency}
/>
```

### Disabled state

```jsx
const [value, setValue] = React.useState('');
const [currency, setCurrency] = React.useState(null);

const currencyOptions = [
    {label: 'RUB', value: 'RUB'},
    {label: 'USD', value: 'USD'},
    {label: 'EUR', value: 'EUR'},
];

<AmountCurrencySelect
    value={value}
    currency={currency}
    currencyOptions={currencyOptions}
    currencyProps={{title: currency ? currency.label : undefined}}
    onChange={setValue}
    onSelect={setCurrency}
    disabled
/>
```

### Error state

```jsx
const [value, setValue] = React.useState('');
const [currency, setCurrency] = React.useState(null);

const currencyOptions = [
    {label: 'RUB', value: 'RUB'},
    {label: 'USD', value: 'USD'},
    {label: 'EUR', value: 'EUR'},
];

<AmountCurrencySelect
    value={value}
    currency={currency}
    currencyOptions={currencyOptions}
    currencyProps={{title: currency ? currency.label : undefined}}
    onChange={setValue}
    onSelect={setCurrency}
    error
/>
```

### Loading state

```jsx
const [value, setValue] = React.useState('');
const [currency, setCurrency] = React.useState(null);

const currencyOptions = [
    {label: 'RUB', value: 'RUB'},
    {label: 'USD', value: 'USD'},
    {label: 'EUR', value: 'EUR'},
];

<AmountCurrencySelect
    value={value}
    currency={currency}
    currencyOptions={currencyOptions}
    currencyProps={{title: currency ? currency.label : undefined}}
    onChange={setValue}
    onSelect={setCurrency}
    loading
/>
```

### Disabled currency

```jsx
const [value, setValue] = React.useState('');
const [currency, setCurrency] = React.useState(null);

const currencyOptions = [
    {label: 'RUB', value: 'RUB'},
    {label: 'USD', value: 'USD'},
    {label: 'EUR', value: 'EUR'},
];

<AmountCurrencySelect
    value={value}
    currency={currency}
    currencyOptions={currencyOptions}
    currencyProps={{title: currency ? currency.label : undefined, disabled: true}}
    onChange={setValue}
    onSelect={setCurrency}
/>
```
