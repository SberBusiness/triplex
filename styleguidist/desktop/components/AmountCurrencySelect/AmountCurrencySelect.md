```jsx
import React, {useState} from 'react';
import {AmountCurrencySelect} from '@sberbusiness/triplex/desktop/components/AmountCurrencySelect/AmountCurrencySelect';
import {ExampleControlPanel} from '../common/ExampleControlPanel/ExampleControlPanel';

const [value, setValue] = useState('');
const [currency, setCurrency] = useState(null);
const [loading, setLoading] = useState(false);
const [error, setError] = useState(false);
const [disabled, setDisabled] = useState(false);
const [currencyDisabled, setCurrencyDisabled] = useState(false);

const currencyOptions = [
    {label: 'RUB', value: 'RUB'},
    {label: 'USD', value: 'USD'},
    {label: 'EUR', value: 'EUR'},
];

const renderExampleControlPanel = () => (
    <ExampleControlPanel>
        <input type="checkbox" checked={loading} onChange={(event) => setLoading(event.target.checked)} data-label="Loading" />
        <input type="checkbox" checked={error} onChange={(event) => setError(event.target.checked)} data-label="Error" />
        <input type="checkbox" checked={disabled} onChange={(event) => setDisabled(event.target.checked)} data-label="Disabled" />
        <input type="checkbox" checked={currencyDisabled} onChange={(event) => setCurrencyDisabled(event.target.checked)} data-label="Disabled (currency)" />
    </ExampleControlPanel>
);

<>
    {renderExampleControlPanel()}  
    <AmountCurrencySelect
        value={value}
        onChange={setValue}
        currency={currency}
        currencyProps={{title: currency ? currency.label : undefined, disabled: currencyDisabled}}
        currencyOptions={currencyOptions}
        onSelect={setCurrency}
        loading={loading}
        error={error}
        disabled={disabled}
    />
</>
```
