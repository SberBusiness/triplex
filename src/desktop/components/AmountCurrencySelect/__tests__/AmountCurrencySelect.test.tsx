import React from 'react';
import renderer from 'react-test-renderer';
import {AmountCurrencySelect} from '../AmountCurrencySelect';
import {allure} from '@jest/unit/allure-report';

jest.mock('@sberbusiness/icons/CurrencySrvIcon20', () => ({
    CurrencySrvIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretdownSrvxIcon16', () => ({
    CaretdownSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/SpinnersmallAniIcon20', () => ({
    SpinnersmallAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/ClosetooltipSrvxIcon16', () => ({
    ClosetooltipSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/triplex/common/utils/uniqueId', () => ({
    uniqueId: jest.fn(() => 0),
}));

describe('AmountCurrencySelect', () => {
    beforeEach(() => {
        allure.feature('AmountCurrencySelect');
    });

    it('renders correctly', () => {
        const component = (
            <AmountCurrencySelect
                value="-8964152,37"
                onChange={() => void 0}
                currency={{label: 'RUB', value: 'RUB'}}
                currencyOptions={[
                    {label: 'RUB', value: 'RUB'},
                    {label: 'USD', value: 'USD'},
                ]}
                onSelect={() => void 0}
            />
        );
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
