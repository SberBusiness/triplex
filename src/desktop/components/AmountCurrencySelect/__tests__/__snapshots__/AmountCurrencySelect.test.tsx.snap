// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountCurrencySelect renders correctly 1`] = `
<div
  className="cssClass[amountCurrencySelect]"
>
  <input
    autoComplete="off"
    className="cssClass[input]"
    onChange={[Function]}
    placeholder="0,00"
    title="-8 964 152,37"
    value="-8 964 152,37"
  />
  <div
    className="cssClass[selectExtended] cssClass[currencySelectExtended]"
  >
    <div
      aria-controls={0}
      aria-expanded={false}
      aria-haspopup="listbox"
      className="cssClass[selectExtendedTarget] cssClass[currencySelectExtendedTarget]"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="combobox"
      tabIndex={0}
    >
      <div
        className="cssClass[label]"
      >
        RUB
      </div>
      <svg
        className="cssClass[caretIcon] hoverable"
      />
    </div>
  </div>
</div>
`;
